src/gui/painting/qpainter.cpp

```diff
@@ -2673,6 +2673,182 @@ QPainterPath QPainter::clipPath() const
+            * placing them directly into a QPainterPath.
+            */
+            QPainterPath path;
+            bool lastWasNothing = true;
+            if (!d->txinv)
+                const_cast<QPainter *>(this)->d_ptr->updateInvMatrix();
+            // ### Falcon: Use QPainterPath
+            for (const QPainterClipInfo &info : qAsConst(d->state->clipInfo)) {
+                switch (info.clipType) {
+
+                    case QPainterClipInfo::PathClip: {
+                        QTransform matrix = (info.matrix * d->invMatrix);
+                        if (lastWasNothing) {
+                            path = info.path * matrix;
+                            lastWasNothing = false;
+                            continue;
+                        }
+                        if (info.operation == Qt::IntersectClip) {
+                            path &= (info.path * matrix);
+                        } else if (info.operation == Qt::NoClip) {
+                            lastWasNothing = true;
+                            path = QPainterPath();
+                        } else {
+                            path = info.path * matrix;
+                        }
+                        break;
+                    }
+                                                 
+                    case QPainterClipInfo::RectClip: {
+                        QTransform matrix = (info.matrix * d->invMatrix);
+                        if (lastWasNothing) {
+                            QPainterPath tempPath;
+                            tempPath.addRect(info.rect);
+                         
+                            path = path + tempPath;
+                            lastWasNothing = false;
+                            continue;
+                        } 
+                        if (info.operation == Qt::IntersectClip) {
+                            QPainterPath tempPath;
+                            tempPath.addRect(info.rect);
+                            tempPath = tempPath * matrix;
+                        
+                            path &= tempPath;
+                        } else if (info.operation == Qt::NoClip) {
+                            lastWasNothing = true;
+                            path = QPainterPath();
+                        } else {
+                            QPainterPath tempPath;
+                            tempPath.addRect(info.rect);
+                            tempPath = tempPath * matrix;
+
+                            path = tempPath;
+                        }
+                        break;
+                    }   
+                    
+                    case QPainterClipInfo::RectFClip: {
+                        QTransform matrix = (info.matrix * d->invMatrix);
+                        if (lastWasNothing) {
+                            QPainterPath tempPath;
+                            tempPath.addRect(info.rectf.toRect());
```

Comment:
You shouldn't need to call `toRect()` throughout here as `info.rectf` should be what you really want. Also, `addRect` only accepts `QRectF`.